Dataflow: Adaptive field flow precision

aschackmull opened the pull request:
This PR builds on top of https://github.com/github/codeql/pull/3880.

This PR introduces adaptive precision of field flow. Before access paths are constructed and used we calculate their expected costs in terms of tuple counts and fan-out. If the set thresholds are exceeded then we use slightly less precision in the access path representation.  The lower precision is still expected to yield sufficient precision for almost all use-cases (in fact, the lowest possible precision used to be the default precision of the library), so this feature is expected to act as a safe-guard against time-outs in degenerate cases.
In particular, https://github.com/github/codeql/pull/3880 makes such degenerate cases much more likely, and with the fix to the semantics of `fieldFlowBranchLimit` that's also included on this PR jdk11 without adaptive precision would experience such a degenerate case: some `AccessPathApprox` would expand to more than 20000 `AccessPath`s applying to hundreds of nodes, which would cost more than 100 million tuples just for the `(Node, AccessPath)` pairs in the final stage.

aschackmull commented:
https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/954/

hvitved commented:
https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/528/

hvitved commented:
> https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/528/

Something is clearly wrong with `mono`; it now takes double time to analyze.

Also, I rebased my PR, so you may want to rebase your PR as well.

aschackmull commented:
Rebased. I also reordered the last two commits (and added an additional sync to make both commits have everything synced). We should test both of these commits to see the effect of the `fieldFlowBranchLimit` fix in isolation as that is probably costly.

aschackmull commented:
All commits: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/956/
Excluding last commit: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/957/

aschackmull commented:
All commits: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/530/
Excluding last commit: https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/531/

hvitved commented:
The pattern is the same for Java and C#: The last commit causes a slow-down of a factor 2 on some projects, so as we agreed offline, we should hold that commit back for now.

hvitved commented on java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -1524,11 +1524,50 @@ private predicate flowCandIsReturned(
-  TConsNil(TypedContent tc, DataFlowType t) { flowCandConsCand(tc, TFrontNil(t), _) } or
+  TConsNil(TypedContent tc, DataFlowType t) {
+    flowCandConsCand(tc, TFrontNil(t), _) and
+    not expensiveLen2unfolding(tc, _)
```
I am not sure why we can't always represent length-1 access paths precisely?

aschackmull replied:
Remember that a length-0 access path can contain a variety of types, so expanding a length-1 access path might also explode. Indeed that's what we check for here.

hvitved replied:
Ah, yes.

hvitved commented on java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -2099,9 +2189,118 @@ private class SummaryCtxSome extends SummaryCtx, TSummaryCtxSome {
+  )
+}
+
+private AccessPathApprox getATail(AccessPathApprox apa, Configuration config) {
```
I would make this a member predicate of `AccessPathApprox`.

aschackmull replied:
Hmm, I'm not sure I agree with that, since it depends on the pruning stage that calculates flow using `AccessPathApprox`, so it's not available at the same stage as the rest of the type and its member predicates.

hvitved replied:
Good point.

hvitved commented on java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -2099,9 +2189,118 @@ private class SummaryCtxSome extends SummaryCtx, TSummaryCtxSome {
+    aps = countPotentialAps(apa, config) and
+    nodes =
+      strictcount(Node n |
+        flow(n, _, _, apa, _)
```
Why not `config` instead of `_`?

aschackmull replied:
Good catch, that should indeed be `config` (not that it's likely to matter much, but still). Similarly `config` ought to be included in `nodeMayUseSummary`.

hvitved replied:
Perhaps the `flow(n, _, _, apa, config)` disjunct should be moved into `nodeMayUseSummary` as it is repeated twice?

hvitved replied:
Indeed, we can move the counting there as well (and rename the predicate).

hvitved reviewed (commented):
LGTM, just a few questions. I have rebased my PR to resolve merge conflicts, so this PR can be rebased as well. And, as we have discussed, we should skip the last commit for now.

hvitved commented:
As we agreed offline, let's get this merged now, and look out for performance regressions:
- https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/544/
- https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1501/
- https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/962/

hvitved commented:
> As we agreed offline, let's get this merged now, and look out for performance regressions:
> 
> * https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/544/
> * https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1501/
> * https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/962/

All look good.
